src/pip/_internal/download.py

```diff
@@ -141,38 +147,141 @@ def user_agent():
     )
 
 
+def _get_keyring_auth(url, username):
+    try:
+        get_credential = keyring.get_credential
+    except AttributeError:
+        logger.debug("Getting password from keyring for %s", url)
+        password = keyring.get_password(url, username)
+        if password:
+            logger.debug("Found password in keyring")
+            return username, password
+    else:
+        logger.debug("Getting credentials from keyring for %s", url)
+        cred = get_credential(url, username)
+        if cred is not None:
+            logger.debug("Found credentials in keyring")
+            return cred.username, cred.password
+
+
 class MultiDomainBasicAuth(AuthBase):
 
-    def __init__(self, prompting=True):
-        # type: (bool) -> None
+    def __init__(self, prompting=True, index_urls=None):
+        # type: (bool, Optional[Values]) -> None
         self.prompting = prompting
+        self.index_urls = index_urls
         self.passwords = {}  # type: Dict[str, AuthInfo]
-
-    def __call__(self, req):
-        parsed = urllib_parse.urlparse(req.url)
-
-        # Split the credentials from the netloc.
-        netloc, url_user_password = split_auth_from_netloc(parsed.netloc)
-
-        # Set the url of the request to the url without any credentials
-        req.url = urllib_parse.urlunparse(parsed[:1] + (netloc,) + parsed[2:])
+        # When the user is prompted to enter credentials and keyring is
+        # available, we will offer to save them. If the user accepts,
+        # this value is set to the credentials they entered. After the
+        # request authenticates, the caller should call
+        # ``save_credentials`` to save these.
+        self._credentials_to_save = None   # type: Tuple[str, str, str]
+
+    def _get_index_url(self, url):
+        """Returns the original index URL matching the requested URL.
+
+        Cached or dynamically generated credentials may work against
+        the original index URL rather than just the netloc.
+
+        The provided url should have had its username and password
+        removed already. If the original index url had credentials then
+        they will be included in the return value.
+
+        Returns None if no matching index was found, or if --no-index
+        was specified by the user.
+        """
+        if not url or not self.index_urls:
+            return None
+
+        for u in self.index_urls:
+            prefix = remove_auth_from_url(u).rstrip("/") + "/"
+            if url.startswith(prefix):
+                return u
+
+    def _get_url_and_credentials(self, original_url,
+                                 allow_netrc=True, allow_keyring=True):
+        """Returns the credentials to use for the provided URL.
```

Comment:
Per PEP 8 / PEP 257, the docstring should begin with the form of a command (so "Return the credentials..."), and ditto for elsewhere in the PR.

Reply:
None of the methods around these changes do this. Want me to update those in the same class for consistency?

Reply:
It's up to you. I was just thinking the ones you were touching, but I'd be okay with fixing the others.

Reply:
I was actually looking at a file in requests when I posted that, so I updated mine. The rest of this class doesn't have docstrings at all :)